StandaloneMmPkg/Library/StandaloneMmCoreEntryPoint/Arm/StandaloneMmCoreEntryPoint.c

```diff
@@ -28,341 +42,1013 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
-    DEBUG ((DEBUG_INFO, "X4 :  0x%x\n", (UINT32)EventCompleteSvcArgs->Arg4));
-    DEBUG ((DEBUG_INFO, "X5 :  0x%x\n", (UINT32)EventCompleteSvcArgs->Arg5));
-    DEBUG ((DEBUG_INFO, "X6 :  0x%x\n", (UINT32)EventCompleteSvcArgs->Arg6));
-    DEBUG ((DEBUG_INFO, "X7 :  0x%x\n", (UINT32)EventCompleteSvcArgs->Arg7));
+  if (CommProtocol == CommProtocolSpmMm) {
+    Status = ValidateSpmMmBootInfo (Arg0, Arg1, Arg2, Arg3);
+  } else if (CommProtocol == CommProtocolFfa) {
+    /*
+     * In case of FF-A, Arg0 is set as FFA_BOOT_INFO address.
+     */
+    Status = ValidateFfaBootInfo (Arg0);
+  }
+
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "Error: Failed to validate boot information!\n"));
+  }
+
+  return Status;
+}
+
+/**
+  Get PHIT hob information from firmware handoff transfer list protocol.
+
+  @param[in]      TlhAddr     Transfer list header address
+
+  @retval         NULL                    Failed to get PHIT hob
+  @retval         Address                 PHIT hob address
+
+**/
+STATIC
+VOID *
+EFIAPI
+GetPhitHobFromTransferList (
+  IN UINTN  TlhAddr
+  )
+{
+  TRANSFER_LIST_HEADER   *Tlh;
+  TRANSFER_ENTRY_HEADER  *Te;
+  VOID                   *HobStart;
 
-    //
-    // ARM TF passes SMC FID of the MM_COMMUNICATE interface as the Event ID upon
-    // receipt of a synchronous MM request. Use the Event ID to distinguish
-    // between synchronous and asynchronous events.
-    //
-    if ((ARM_SMC_ID_MM_COMMUNICATE != (UINT32)EventCompleteSvcArgs->Arg0) &&
-        (ARM_SVC_ID_FFA_MSG_SEND_DIRECT_REQ != (UINT32)EventCompleteSvcArgs->Arg0))
+  Tlh = (TRANSFER_LIST_HEADER *)TlhAddr;
+
+  Te = TlFindFirstEntry (Tlh, TRANSFER_ENTRY_TAG_ID_HOB_LIST);
+  if (Te == NULL) {
+    DEBUG ((DEBUG_ERROR, "Error: No Phit hob is present in transfer list...\n"));
+
+    return NULL;
+  }
+
+  HobStart = TlGetEntryData (Te);
+
+  return HobStart;
+}
+
+/**
+  Get PHIT hob information from FF-A boot information.
+
+  @param[in]      FfaBootInfoHeaderAddr   FF-A boot information header address
+
+  @retval         NULL                    Failed to get PHIT hob
+  @retval         Address                 PHIT hob address
+
+**/
+STATIC
+VOID *
+EFIAPI
+GetPhitHobFromFfaBootInfo (
+  IN UINTN  FfaBootInfoHeaderAddr
+  )
+{
+  EFI_FFA_BOOT_INFO_HEADER  *FfaBootInfoHeader;
+  EFI_FFA_BOOT_INFO_DESC    *FfaBootInfoDesc;
+  UINT32                    Idx;
+
+  FfaBootInfoHeader = (EFI_FFA_BOOT_INFO_HEADER *)FfaBootInfoHeaderAddr;
+
+  for (Idx = 0; Idx  < FfaBootInfoHeader->CountBootInfoDesc; Idx++) {
+    FfaBootInfoDesc = (EFI_FFA_BOOT_INFO_DESC *)((FfaBootInfoHeaderAddr +
+                                                  FfaBootInfoHeader->OffsetBootInfoDesc) +
+                                                 (Idx * FfaBootInfoHeader->SizeBootInfoDesc));
+
+    if ((FFA_BOOT_INFO_TYPE (FfaBootInfoDesc->Type) != FFA_BOOT_INFO_TYPE_STD) ||
+        (FFA_BOOT_INFO_TYPE_ID (FfaBootInfoDesc->Type) != FFA_BOOT_INFO_TYPE_ID_HOB) ||

```

Comment:
If there is a SPMC such as hafnium in the middle, will the TFA created HOB list still be able to propagate through? Or the boot info needs to be recreated by SPMC?

I think I am seeing some errors `No Phit hob is present in FfaBootInfo` when having the hafnium.

Reply:
> Or the boot info needs to be recreated by SPMC
Maybe this way. 

> I think I am seeing some errors No Phit hob is present in FfaBootInfo when having the hafnium.
That's way someone is preparing the internal patch for hafnium to pass HOB List via FF-A boot protocol IIRC.

Anyway, In case of StandaloneMm, If it's not delivered Hob list, It couldn't initialize properly.